Source/CesiumRuntime/Private/CesiumGltfPrimitiveComponent.h

```diff
@@ -105,16 +106,118 @@ class UCesiumGltfPrimitiveComponent : public UStaticMeshComponent {
+        ETeleportType::TeleportPhysics);
+  } else {
+    // Unreal will yell at us for calling SetRelativeTransform on a static
+    // object, but we still need to adjust (accurately!) for origin rebasing
+    // and georeference changes. It's "ok" to move a static object in this way
+    // because, we assume, the globe and globe-oriented lights, etc. are
+    // moving too, so in a relative sense the object isn't actually moving.
+    // This isn't a perfect assumption, of course.
+    uobject->SetRelativeTransform_Direct(transform);
+    uobject->UpdateComponentToWorld();
+    uobject->MarkRenderTransformDirty();
+    uobject->SendPhysicsTransform(ETeleportType::ResetPhysics);
+  }
+}
+
+template <typename UnrealType>
+void BeginDestroyPrimitive(
+    UnrealType* uobject,
+    CesiumGltfPrimitiveBase* cesiumData) {
+  // Clear everything we can in order to reduce memory usage, because this
+  // UObject might not actually get deleted by the garbage collector until
+  // much later.
+  cesiumData->BeginDestroyPrimitive();
+  UMaterialInstanceDynamic* pMaterial =
+      Cast<UMaterialInstanceDynamic>(uobject->GetMaterial(0));
+  if (pMaterial) {
+    CesiumLifetime::destroy(pMaterial);
+  }
+
+  UStaticMesh* pMesh = uobject->GetStaticMesh();
+  if (pMesh) {
+    UBodySetup* pBodySetup = pMesh->GetBodySetup();
+
+    if (pBodySetup) {
+      CesiumLifetime::destroy(pBodySetup);
+    }
+
+    CesiumLifetime::destroy(pMesh);
+  }
+}
+
+template <typename T>
+class CesiumGltfPrimitive : public CesiumGltfPrimitiveBase {
+public:
+  T* _uobject;
+};
+
+UCLASS()
+class UCesiumGltfPrimitiveComponent : public UStaticMeshComponent {
+  GENERATED_BODY()
+
+public:
+  // Sets default values for this component's properties
+  UCesiumGltfPrimitiveComponent();
+  virtual ~UCesiumGltfPrimitiveComponent();
+
+  virtual void BeginDestroy() override;
+
+  virtual FBoxSphereBounds CalcBounds(const FTransform& LocalToWorld) const;
+
+  using UPrimitiveComponent::SendPhysicsTransform;
+
+  friend CesiumGltfPrimitiveBase* getPrimitiveBase(USceneComponent* Component);
+  friend const CesiumGltfPrimitiveBase*
+  getPrimitiveBase(const USceneComponent* Component);
+
+protected:
+  CesiumGltfPrimitiveBase _cesiumBase;
+};
+
+UCLASS()
+class UCesiumGltfInstancedComponent : public UInstancedStaticMeshComponent {
+  GENERATED_BODY()
+
+public:
+  // Sets default values for this component's properties
+  UCesiumGltfInstancedComponent();
+  virtual ~UCesiumGltfInstancedComponent();
 
   virtual void BeginDestroy() override;
 
   virtual FBoxSphereBounds CalcBounds(const FTransform& LocalToWorld) const;
+
+  using UPrimitiveComponent::SendPhysicsTransform;
+
+  friend CesiumGltfPrimitiveBase* getPrimitiveBase(USceneComponent* Component);
+  friend const CesiumGltfPrimitiveBase*
+  getPrimitiveBase(const USceneComponent* Component);
+
+protected:
+  CesiumGltfPrimitiveBase _cesiumBase;
 };
+
+CesiumGltfPrimitiveBase* getPrimitiveBase(USceneComponent* pComponent);
```

Comment:
This should be a virtual method on an interface (perhaps `ICesiumTile`) that is implemented by both primitive types, rather than a global function that uses a dynamic cast. See here for how to create interfaces in UE:
https://dev.epicgames.com/documentation/en-us/unreal-engine/interfaces-in-unreal-engine?application_version=5.2

Reply:
Thinks for that link, it's very helpful. I would note that its example shows inheriting from the interface via multiple inheritance, but that would be unproductive 😎

Reply:
>  I could note that its example shows inheriting from the interface via multiple inheritance

Sure, that's the only way to implement interfaces in C++. Multiple inheritance is widely considered acceptable when it's inheritance of an interface only (no implementation, no fields). Inheriting from multiple classes with fields has extra gotchas.

This goes back to Effective C++, item 40 if my quick google search is to be believed (I don't have the book handy).

As always with these sorts of rules of thumb ("don't use multiple inheritance, unless all but one is an interface"), there are exceptions. But this isn't one of them.